File: python/282_ExpressionAddOperators.py
from unittest import TestCase
# https://leetcode.com/problems/expression-add-operators/


class ExpressionAddOperators(TestCase):
    def addOperators(self, num, target):
        """
        :type num: str
        :type target: int
        :rtype: List[str]
        """
        n = len(num)
        ret = []
        def dfs(i, v, last, exp):
            if i == n:
                if v+last == target: ret.append(exp)
                return
            for j in range(i+1, n+1):
                if num[i] == '0' and j-i > 1:
                    break
                cur = num[i:j]
                vcur = int(cur)

                if i == 0:
                    dfs(j, 0, vcur, cur)
                else:
                    dfs(j, v+last, vcur, exp+"+"+cur)
                    dfs(j, v+last, -vcur, exp+"-"+cur)
                    dfs(j, v, last*vcur, exp+"*"+cur)
        dfs(0, 0, 0, "")
        return ret

    def addOperators(self, num, target):
        n, res = len(num), []
        def dfs(i, sum, mul, s, exp):
            if i == n:
                if target == sum + mul*int(s): res.append(exp)
                return
            c = num[i]
            if s != "0": dfs(i+1, sum, mul, s+c, exp+c)
            mul *= int(s)
            dfs(i+1, sum, mul, c, exp+'*'+c)
            sum += mul
            dfs(i+1, sum, 1, c, exp+'+'+c)
            dfs(i+1, sum, -1, c, exp+'-'+c)
        if n: dfs(1, 0, 1, num[0], num[0])
        return res

    def test1(self):
        self.assertEqual({"1+2+3", "1*2*3"}, set(self.addOperators("123", 6)))

    def test2(self):
        self.assertEqual({"2*3+2", "2+3*2"}, set(self.addOperators("232", 8)))

    def test3(self):
        self.assertEqual({"1*0+5", "10-5"}, set(self.addOperators("105", 5)))

    def test4(self):
        self.assertEqual({"0+0", "0-0", "0*0"}, set(self.addOperators("00", 0)))

    def test5(self):
        self.assertEqual(set(), set(self.addOperators("3456237490", 9191)))

File: kotlin/src/main/kotlin/282.ExpressionAddOperators.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/expression-add-operators/


class ExpressionAddOperators{
  fun addOperators(num: String, target: Int): List<String> {
    val n = num.length
    val ret = mutableListOf<String>()

    fun dfs(i: Int, v: Long, last: Long, exp: String){
      if(i == n){
        if(v+last == target.toLong()) ret.add(exp)
        return
      }
      for(j in i+1..n){
        if(num[i] == '0' && j-i > 1) break

        val cur = num.substring(i, j)
        val curv = cur.toBigInteger().toLong()
        if(i == 0)
          dfs(j, v, curv, cur)
        else{
          dfs(j, v+last, curv, "$exp+$cur")
          dfs(j, v+last, -curv, "$exp-$cur")
          dfs(j, v, last*curv, "$exp*$cur")
        }
      }
    }
    dfs(0, 0, 0, "")
    return ret
  }

  @Test
  fun test1(){
    assertEquals(setOf("1+2+3", "1*2*3"), addOperators("123", 6).toSet())
  }

  @Test
  fun test2(){
    assertEquals(setOf("2*3+2", "2+3*2"), addOperators("232", 8).toSet())
  }

  @Test
  fun test3(){
    assertEquals(setOf("1*0+5", "10-5"), addOperators("105", 5).toSet())
  }

  @Test
  fun test4(){
    assertEquals(setOf("0+0", "0-0", "0*0"), addOperators("00", 0).toSet())
  }

  @Test
  fun test5(){
    assertEquals(setOf<String>(), addOperators("3456237490", 9191).toSet())
  }

}